src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ClientService.java

```diff
@@ -439,4 +464,121 @@ public ClientType deleteTlsCertificate(ClientId id, String certificateHash)
+     * @throws AdditionalMemberAlreadyExistsException if tried to add a new member, and
+     * security server already has owner member + one additional member
+     * @throws UnhandledWarningsException if tried to add client associated with a member which
+     * does not exist in global conf yet, and ignoreWarnings was false
+     */
+    public ClientType addLocalClient(String memberClass,
+            String memberCode,
+            String subsystemCode,
+            IsAuthentication isAuthentication,
+            boolean ignoreWarnings) throws ClientAlreadyExistsException,
+            AdditionalMemberAlreadyExistsException, UnhandledWarningsException {
+
+        ClientId clientId = ClientId.create(globalConfFacade.getInstanceIdentifier(),
+                memberClass,
+                memberCode,
+                subsystemCode);
+
+        ClientType existingLocalClient = getLocalClient(clientId);
+        ClientId ownerId = serverConfService.getSecurityServerOwnerId();
+        if (existingLocalClient != null) {
+            throw new ClientAlreadyExistsException("client " + clientId + " already exists");
+        }
+        if (clientId.getSubsystemCode() == null) {
+            // adding member - check that we dont already have owner + one additional member
+            List<ClientType> existingMembers = getAllLocalMembers();
+            Optional<ClientType> additionalMember = existingMembers.stream()
+                    .filter(m -> !ownerId.equals(m.getIdentifier()))
+                    .findFirst();
+            if (additionalMember.isPresent()) {
+                throw new AdditionalMemberAlreadyExistsException("additional member "
+                        + additionalMember.get().getIdentifier() + " already exists");
+            }
+        }
+
+        // check if the member associated with clientId exists in global conf
+        ClientId memberId = clientId.getMemberId();
+        if (globalConfFacade.getMemberName(memberId) == null) {
+            // unregistered member
+            if (!ignoreWarnings) {
+                WarningDeviation warning = new WarningDeviation(WARNING_UNREGISTERED_MEMBER, memberId.toShortString());
+                throw new UnhandledWarningsException(warning);
+            }
+        }
+
+        boolean clientRegistered = globalConfService.isSecurityServerClientForThisInstance(clientId);
+        ClientType client = new ClientType();
+        client.setIdentifier(getPossiblyManagedEntity(clientId));
+        if (clientRegistered) {
+            client.setClientStatus(ClientType.STATUS_REGISTERED);
+        } else {
+            client.setClientStatus(ClientType.STATUS_SAVED);
+        }
+        client.setIsAuthentication(isAuthentication.name());
+        ServerConfType serverConfType = serverConfService.getServerConf();
+        client.setConf(serverConfType);
+        serverConfType.getClient().add(client);
+
+        clientRepository.saveOrUpdate(client);
+        return client;
```

Comment:
Just wondering why this does not require saving the `serverConfType` itself? Since there is `cascade=none` from `ClientType` to `ServerConf` now.

Reply:
Answering to myself here: probably the `@Transactional` annotation kicking in

Reply:
`serverConfType` is already a managed entity; if there was some change that needs to be saved to database, it would automatically be saved by the JPA implementation when `flush` happens (probably, but not guaranteed, when transaction commits, so as you said, when @Transactional kicks in). You only need to manually save entities when they are new or transient.

Also serverConfType itself is not changed, all the DB changes are related to client object and client table.

This is probably a bit inaccurate / fuzzy explanation, tried to find a more exact and concrete explanation but did not find one quickly, sorry.